File: packages/sandboxd/src/process/output.rs
//! Bounded stdout/stderr capture for runtime client processes.
//!
//! Process start and health errors include recent output so callers get useful
//! diagnostics without retaining unbounded child logs in memory.

use std::collections::VecDeque;
use std::io::Read;
use std::thread::{self, JoinHandle};

use crate::process::*;

impl ProcessOutputCapture {
    pub(super) fn new() -> Self {
        Self {
            stdout: Arc::new(Mutex::new(TailBuffer::new(4 * 1024))),
            stderr: Arc::new(Mutex::new(TailBuffer::new(8 * 1024))),
            capture_threads: Arc::new(Mutex::new(Vec::new())),
        }
    }

    pub(super) fn record_stdout(&self, bytes: &[u8]) {
        self.stdout
            .lock()
            .expect("stdout tail buffer lock should not be poisoned")
            .append(bytes);
    }

    pub(super) fn record_stderr(&self, bytes: &[u8]) {
        self.stderr
            .lock()
            .expect("stderr tail buffer lock should not be poisoned")
            .append(bytes);
    }

    pub(super) fn stdout_tail(&self) -> Option<String> {
        self.stdout
            .lock()
            .expect("stdout tail buffer lock should not be poisoned")
            .snapshot()
    }

    pub(super) fn stderr_tail(&self) -> Option<String> {
        self.stderr
            .lock()
            .expect("stderr tail buffer lock should not be poisoned")
            .snapshot()
    }

    pub(super) fn register_capture_thread(&self, capture_thread: JoinHandle<()>) {
        self.capture_threads
            .lock()
            .expect("capture thread list lock should not be poisoned")
            .push(capture_thread);
    }

    pub(super) fn finish_capture_threads(&self) {
        let capture_threads = {
            let mut capture_threads = self
                .capture_threads
                .lock()
                .expect("capture thread list lock should not be poisoned");
            std::mem::take(&mut *capture_threads)
        };

        for capture_thread in capture_threads {
            let _ = capture_thread.join();
        }
    }

    pub(super) fn collect_tails_after_process_exit(&self) -> ProcessOutputTails {
        self.finish_capture_threads();

        let stdout_tail = self.stdout_tail();
        let stderr_tail = self.stderr_tail();
        ProcessOutputTails {
            stdout_captured: stdout_tail.is_some(),
            stderr_captured: stderr_tail.is_some(),
            stdout_tail,
            stderr_tail,
        }
    }
}

impl TailBuffer {
    pub(super) fn new(max_bytes: usize) -> Self {
        Self {
            max_bytes,
            bytes: VecDeque::new(),
        }
    }

    pub(super) fn append(&mut self, chunk: &[u8]) {
        for &byte in chunk {
            self.bytes.push_back(byte);
        }
        while self.bytes.len() > self.max_bytes {
            let _ = self.bytes.pop_front();
        }
    }

    pub(super) fn snapshot(&self) -> Option<String> {
        if self.bytes.is_empty() {
            return None;
        }

        let bytes = self.bytes.iter().copied().collect::<Vec<_>>();
        Some(String::from_utf8_lossy(&bytes).into_owned())
    }
}

pub(super) fn spawn_output_capture_thread<R>(
    mut reader: R,
    output_capture: ProcessOutputCapture,
    stream: OutputStream,
) -> JoinHandle<()>
where
    R: Read + Send + 'static,
{
    thread::spawn(move || {
        let mut buffer = [0_u8; 4096];
        loop {
            match reader.read(&mut buffer) {
                Ok(0) => break,
                Ok(bytes_read) => match stream {
                    OutputStream::Stdout => output_capture.record_stdout(&buffer[..bytes_read]),
                    OutputStream::Stderr => output_capture.record_stderr(&buffer[..bytes_read]),
                },
                Err(_) => break,
            }
        }
    })
}

File: packages/sandboxd/process/output.go
package process

import (
	"io"
	"sync"
)

const (
	DefaultProcessStdoutTailBytes = 4 * 1024
	DefaultProcessStderrTailBytes = 8 * 1024
)

type OutputStream string

const (
	OutputStreamStdout OutputStream = "stdout"
	OutputStreamStderr OutputStream = "stderr"
)

type OutputTails struct {
	StdoutTail     *string
	StderrTail     *string
	StdoutCaptured bool
	StderrCaptured bool
}

type OutputCapture struct {
	stdout         *TailBuffer
	stderr         *TailBuffer
	stdoutMutex    sync.Mutex
	stderrMutex    sync.Mutex
	captureWaiters sync.WaitGroup
}

func NewOutputCapture() *OutputCapture {
	return &OutputCapture{
		stdout: NewTailBuffer(DefaultProcessStdoutTailBytes),
		stderr: NewTailBuffer(DefaultProcessStderrTailBytes),
	}
}

func (capture *OutputCapture) RecordStdout(bytes []byte) {
	capture.stdoutMutex.Lock()
	defer capture.stdoutMutex.Unlock()
	capture.stdout.Append(bytes)
}

func (capture *OutputCapture) RecordStderr(bytes []byte) {
	capture.stderrMutex.Lock()
	defer capture.stderrMutex.Unlock()
	capture.stderr.Append(bytes)
}

func (capture *OutputCapture) StdoutTail() *string {
	capture.stdoutMutex.Lock()
	defer capture.stdoutMutex.Unlock()
	return capture.stdout.Snapshot()
}

func (capture *OutputCapture) StderrTail() *string {
	capture.stderrMutex.Lock()
	defer capture.stderrMutex.Unlock()
	return capture.stderr.Snapshot()
}

func (capture *OutputCapture) RegisterCaptureReader(reader io.Reader, stream OutputStream) {
	capture.captureWaiters.Add(1)
	go func() {
		defer capture.captureWaiters.Done()
		CaptureReaderOutput(reader, capture, stream)
	}()
}

func (capture *OutputCapture) FinishCaptureReaders() {
	capture.captureWaiters.Wait()
}

func (capture *OutputCapture) CollectTailsAfterProcessExit() OutputTails {
	capture.FinishCaptureReaders()
	stdoutTail := capture.StdoutTail()
	stderrTail := capture.StderrTail()
	return OutputTails{
		StdoutCaptured: stdoutTail != nil,
		StderrCaptured: stderrTail != nil,
		StdoutTail:     stdoutTail,
		StderrTail:     stderrTail,
	}
}

type TailBuffer struct {
	maxBytes int
	bytes    []byte
}

func NewTailBuffer(maxBytes int) *TailBuffer {
	return &TailBuffer{maxBytes: maxBytes}
}

func (buffer *TailBuffer) Append(chunk []byte) {
	buffer.bytes = append(buffer.bytes, chunk...)
	if len(buffer.bytes) <= buffer.maxBytes {
		return
	}
	buffer.bytes = append([]byte(nil), buffer.bytes[len(buffer.bytes)-buffer.maxBytes:]...)
}

func (buffer *TailBuffer) Snapshot() *string {
	if len(buffer.bytes) == 0 {
		return nil
	}
	snapshot := string(buffer.bytes)
	return &snapshot
}

func CaptureReaderOutput(reader io.Reader, outputCapture *OutputCapture, stream OutputStream) {
	buffer := make([]byte, 4096)
	for {
		bytesRead, err := reader.Read(buffer)
		if bytesRead > 0 {
			switch stream {
			case OutputStreamStdout:
				outputCapture.RecordStdout(buffer[:bytesRead])
			case OutputStreamStderr:
				outputCapture.RecordStderr(buffer[:bytesRead])
			}
		}
		if err != nil {
			return
		}
	}
}
